### PR TITLE
Integration Base Testing

### PR DESCRIPTION
**Pull request by dehume**
Adding integration tests for the terraform provider using the `materialized`. This creates a simple docker compose of the `materialized` and a container with the provider installed locally. It then keeps the provider container running (don't love) and then runs a series of individual terraform commands for specific resources using `-target`.

### CODE REVIEW

**Comment by dehume** [comment]
@bobbyiliev This works locally. Trying to figure out what the Terraform error could be.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+version: "3.9"
+
+services:
+
+  materialized:
+    image: materialize/materialized:latest
+    container_name: materialized
+    ports:
+      - 6875:6875
+    healthcheck: {test: curl -f localhost:6875, interval: 1s, start_period: 30s}
```
Comment (bobbyiliev) [inline docker-compose.yml:11]: I believe that the problem is with the healthcheck.
Basically it takes ~10s or so for the materialized container to be fully ready, but only using the `depends_on` directive, would wait for just a second or so.



```suggestion
      - 6878:6878
    healthcheck: {test: curl -f localhost:6878/api/readyz, interval: 1s, start_period: 30s}
```

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  provider:
+    build:
+      context: .
+    container_name: provider
+    depends_on:
+      - materialized
```
Comment (bobbyiliev) [inline docker-compose.yml:18]:
```suggestion
      materialized: {condition: service_healthy}
 ```

File: docker-compose.yml (hunk 1 above)
Comment (bobbyiliev) [reply docker-compose.yml:11]: I've managed to reproduce the failure locally, after running the test a few times in a row until I hit the same error. My guess is that locally the materialized container becomes healthy quicker so we don't hit the problem as often.

**Review by bobbyiliev — CHANGES_REQUESTED** [review]

**Comment by dehume** [comment]
@philip-stoev wanted to get your thoughts on the layout of our integration tests for the Terraform provider. Part of me was thinking we could have individual Terraform commands to spin and down various resources but I could also be convinced that it would be better to just maintain a single stack and spin up and down everything at once. If there are any patterns we use elsewhere, I'm happy to follow those.

**Comment by dehume** [comment]
Setting up integration tests for certain issues and starting to uncover some issues with the more complicated resources. Moving those to separate issues:
* #22 
* #23 
* #24
We may want to add additional services to the docker compose (kafka, postgres...) to handle these. Looking at localstack for AWS Privatelink but there is no localstack option for that.

**Comment by benesch** [comment]
> Looking at localstack for AWS Privatelink but there is no localstack option for that.

That's okay! You can still _create_ the connection in Materialize, with a dummy service name. It just won't connect to anything. That should be sufficient for integration testing the provider. We have a separate end to end test elsewhere that will make sure that AWS PrivateLink connections _actually_ work end to end with real AWS VPC endpoints, etc.

**Review by bobbyiliev — APPROVED** [review]

**Comment by philip-stoev** [comment]
@dehume I have zero experience with Terraform, but will try to provide feedback by tomorrow. Please feel free to not wait for me though.

**Pull request merged**